crates/uopool/src/mempool.rs

```diff
@@ -30,6 +30,7 @@ pub trait Mempool: Debug {
     fn get(&self, uo_hash: &UserOperationHash) -> Result<Option<UserOperation>, Self::Error>;
     fn get_all_by_sender(&self, addr: &Address) -> Self::UserOperations;
     fn get_number_by_sender(&self, addr: &Address) -> usize;
+    fn get_prev_by_sender(&self, uo: &UserOperation) -> Option<UserOperation>;
```

Comment:
This method looks identical for two implementations. We could define a `MempoolExtension` trait and define this as a extension method. In that case, you don't have to replicate identical implementations.

Reply:
https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html here is some links for reference.

The idea is to add methods to an existing type. 
We could define
```rust
trait MempoolExt{
    fn get_prev_by_sender(&self, uo: &UserOperation) -> Option<UserOperation>;
}

impl<T: Mempool> MempoolExt for T {
    fn get_prev_by_sender(&self, uo: &UserOperation) -> Option<UserOperation> {
        self.get_all_by_sender(&uo.sender)
            .into_iter()
            .filter(|prev_uo| prev_uo.nonce == uo.nonce - 1)
            .max_by_key(|prev_uo| prev_uo.max_priority_fee_per_gas)
    }
}
```

Then you could directly use `get_prev_by_sender` for all the implementation of Mempool when you import the package with `use uopool::MempoolExt`. In this case, you don't have to replicate the same implementation twice but only one time.

Reply:
Actually there is another easier way to do that without defining another trait. Because the `Mempool` trait is actually defined within the our crates, We could just 
```rust
pub trait Mempool: Debug {
    type UserOperations: IntoIterator<Item = UserOperation>;
    type CodeHashes: IntoIterator<Item = CodeHash>;
    type Error;
    fn add(
        &mut self,
        uo: UserOperation,
        ep: &Address,
        chain_id: &U256,
    ) -> Result<UserOperationHash, Self::Error>;
    fn get(&self, uo_hash: &UserOperationHash) -> Result<Option<UserOperation>, Self::Error>;
    fn get_all_by_sender(&self, addr: &Address) -> Self::UserOperations;
    fn get_number_by_sender(&self, addr: &Address) -> usize;
    fn get_prev_by_sender(&self, uo: &UserOperation) -> Option<UserOperation>{
        self.get_all_by_sender(&uo.sender)
            .into_iter()
            .filter(|prev_uo| prev_uo.nonce == uo.nonce - 1)
            .max_by_key(|prev_uo| prev_uo.max_priority_fee_per_gas)
    }
    fn has_code_hashes(&self, uo_hash: &UserOperationHash) -> Result<bool, Self::Error>;
    fn set_code_hashes(
        &mut self,
        uo_hash: &UserOperationHash,
        hashes: &Self::CodeHashes,
    ) -> Result<(), Self::Error>;
    fn get_code_hashes(&self, uo_hash: &UserOperationHash) -> Self::CodeHashes;
    fn remove(&mut self, uo_hash: &UserOperationHash) -> Result<(), Self::Error>;
    // Get UserOperations sorted by max_priority_fee_per_gas without dup sender
    fn get_sorted(&self) -> Result<Self::UserOperations, Self::Error>;
    fn get_all(&self) -> Self::UserOperations;
    fn clear(&mut self);
}
```

Reply:
These are 2 ways to keep DRY. Extension method make extend external trait possible but in our case the trait is not external so the second way could work too.

Reply:
Ok, got it. I decided to go for second option.